src/main/java/com/example/demo/service/book/BookServiceImpl.java

```diff
@@ -62,4 +78,18 @@ public List<BookDto> search(BookSearchParameters params, Pageable pageable) {
                 .map(bookMapper::toDto)
                 .toList();
     }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> getAllByCategoryId(Long categoryId, Pageable pageable) {
+        return bookRepository.findAllByCategoryId(categoryId, pageable).stream()
+                .map(bookMapper::toDtoWithoutCategories)
+                .toList();
+    }
+
+    public Set<Category> mapCategoryIdsToCategories(List<Long> categoryIds) {
+        return categoryIds.stream()
+                .map(categoryId -> categoryRepository.findById(categoryId)
+                        .orElseThrow(() -> new EntityNotFoundException("Category not found")))
```

Comment:
provide request parameter

Reply:
added @PathVariable("id") because the method  in the controller according to the task condition is public List getBooksByCategoryId(Long id) (endpoint: "/{id}/books")